### PR TITLE
Scheduling primitives: startOn and completeOn

### PR DESCRIPTION
**Pull request by gavrix**

### CODE REVIEW

File: Promise/PromiseExtension.swift [hunk 1]
```diff
@@ -40,12 +76,12 @@ extension Promise {
   ///   - queue: queue to schedule `dispatchAfter` on
   /// - Returns: new `Promise` instance.
   
-  public func delayed(for delay: DispatchTimeInterval, on queue: DispatchQueue = DispatchQueue.main) -> Promise {
+  public func delayed(for delay: DispatchTimeInterval, on queue: DispatchQueueType = DispatchQueue.main) -> Promise {
```
Comment (sav007) [inline Promise/PromiseExtension.swift]: Should it be called `completeOn(queue: DispatchQueueType)` to be symmetric with `startOn` ?

File: Promise/PromiseExtension.swift [hunk 2]
```diff
@@ -8,6 +8,21 @@
 
 import Foundation
 
+public protocol DispatchQueueType {
+  func async(execute work: DispatchWorkItem)
+  func asyncAfter(deadline: DispatchTime, execute work: DispatchWorkItem)
```
Comment (sav007) [inline Promise/PromiseExtension.swift]: nit: `deadline` -> `time`?

File: Promise/PromiseExtension.swift (hunk 1 above)
Comment (sav007) [reply Promise/PromiseExtension.swift]: Oh we have this? So not sure do we really need this extension?

File: Promise/PromiseExtension.swift (hunk 2 above)
Comment (gavrix) [reply Promise/PromiseExtension.swift]: This has to match the interface that `DispatchQueue` already has. I'm extracting protocol from existing type.

File: Promise/PromiseExtension.swift (hunk 1 above)
Comment (gavrix) [reply Promise/PromiseExtension.swift]: Agree, `delayed` should not be part of a core framework, probably. Let's make another PR to remove it?

File: PromiseTests/TestPromiseScheduling.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+        let p: Promise<Int, NoError> = Promise { resolver in
+            // see if onStarted action has been already been executed.
+            XCTAssertTrue(onExecuted)
+            resolver.resolve(with: 42)
+        }
+        
+        let exp = expectation(description: "")
+        p.startOn(queue: observableQueue).whenComplete { _ in exp.fulfill() }
+        waitForExpectations(timeout: 1.0)
```
Comment (sav007) [inline PromiseTests/TestPromiseScheduling.swift]: am I wrong but this is unnecessary as all tasks are executed in main? so we can have empty whenComplete?

File: Promise/PromiseExtension.swift [hunk 4]
```diff
@@ -8,6 +8,21 @@
...
+
+extension DispatchQueue: DispatchQueueType {
+//  public func async(execute work: @escaping () -> Void) {
+//    self.async(execute: work)
+//  }
+//  
+//  public func asyncAfter(deadline: DispatchTime, execute work: @escaping () -> Void) {
+//    self.asyncAfter(deadline: deadline, execute: work)
+//  }
+}
```
Comment (sav007) [inline Promise/PromiseExtension.swift]: What the intention of this? and why it's commented out?

Comment (gavrix) [reply Promise/PromiseExtension.swift]: So, we need mark `DispatchQueue` conforming to `DispatchQueueType` but it should be empty — we already have all the implementations (in fact, we draw interface for `DispatchQueueType` from what DispatchQueue already have). But comments are — I'll remove them. Good cach! 👍

File: PromiseTests/TestPromiseScheduling.swift (hunk 3 above)
Comment (gavrix) [reply PromiseTests/TestPromiseScheduling.swift]: `startOn` will perform start func asynchronouslyon given queue. If  you want to wait till that async call is complete in this function, you have to setup `expectation`.

**Pull request merged**